cmd/clusterctl/client/cluster/cert_manager.go

```diff
@@ -94,32 +91,34 @@ func (cm *certManagerClient) Images() ([]string, error) {
 	return images, nil
 }
 
-// EnsureWebhook makes sure the cert-manager Web-hook is Available in a cluster:
-// this is a requirement to install a new provider
+// EnsureCertManager makes sure cert-manager is running and its API is available.
+// This is required to install a new provider.
 // Nb. In order to provide a simpler out-of-the box experience, the cert-manager manifest
 // is embedded in the clusterctl binary.
-func (cm *certManagerClient) EnsureWebhook() error {
+func (cm *certManagerClient) EnsureCertManager() error {
 	log := logf.Log
 
-	// Checks if the cert-manager web-hook already exists, if yes, exit immediately
-	hasWebhook, err := cm.hasWebhook()
+	// Gets the cert-manager objects from the embedded assets.
+	objs, err := cm.getManifestObjs()
 	if err != nil {
 		return err
 	}
-	if hasWebhook {
-		return nil
-	}
 
-	// Otherwise install cert-manager
-	log.Info("Installing cert-manager")
-
-	// Gets the cert-manager objects from the embedded assets.
-	objs, err := cm.getManifestObjs()
+	// Skip re-installing cert-manager if all objects already exist
+	allObjectsExist, err := cm.allObjectsExist(ctx, log, objs)
 	if err != nil {
-		return nil
+		return err
 	}
+	// Perform a rough check to see if all cert-manager resources exist.
+	if allObjectsExist {
+		log.Info("Skipping installing cert-manager as it is already installed")
+		// Wait for the cert-manager API to be ready, in case it is not already.
+		return cm.waitForAPIReady(ctx, log)
+	}
+
+	log.Info("Installing cert-manager")
 
-	// installs the web-hook
+	// Install all cert-manager manifests
```

Comment:
Question. Should we start to add a version label to each component so we can properly plan for cert-manager upgrades?

Reply:
This definitely seems sensible - but we will need to store that version number somewhere, and remember to update it if the YAML asset changes. I wonder if we can have an automated test for this..? Maybe even just take a hash of the YAML file and have a variable: `versionNumber` which is a tuple of the version number and the hash of that file?

That way if you change the file, the hash will no longer match, forcing people to look at that area of code?

Perhaps you have other instances of this somewhere and have an existing pattern to solve it? 😄

Reply:
Done in https://github.com/kubernetes-sigs/cluster-api/pull/3313/commits/4aa658b1ff158452d4987c1727985357c33a1fe2 - keen to hear your comments on the technique for keeping the version number up to date.

Reply:
I personally like the idea of the hash, so we can manage adjustment to the manifest without bumping a release.

Reply:
I personally like the idea of the hash, so we can manage adjustment to the manifest without bumping a release.